[WIP] Add mediator support

chimp1984 opened the pull request:
This is heavily work in progress but I thought it might be useful for other developers to share the development progress.

For the new trade protocol we need a mediator which will be very similar to arbitration just that mediators do not have a 3rd key and cannot do a payout but only give the traders a recommendation for the payout. If both traders agree to that result they can complete the trade by both signing the payout tx. Only if they do not find consensus they need to go to the next step which is currently the arbitration. 
So before traders can open a dispute to arbitrators they need to go through the mediation first and only if both do not agree to the mediators decision arbitration will be used.

This should reduce arbitration cases a lot and is a necessary feature for the new trade protocol but can be deployed as addition with the current protocol.

To make sure the feature is available for both traders (both have updated) we can use an activation date and set a mandatory version number after a few days after release, so we enforce traders to update and enable that feature. Otherwise it would be complicated if one peer has updated but the other not...

chimp1984 committed:
Refactoring: Move arbitration package inside dispute package

chimp1984 committed:
Use abstract base class DisputeResolver for arbitrator

chimp1984 committed:
Refactoring: Move mediator to mediator package.

chimp1984 committed:
Let Mediator inherit DisputeResolver.

chimp1984 committed:
Do not use protobuf inheritance

- Do not use protobuf inheritance for Arbitrator and Mediator as it
would break backward compatibility (and protobuf inheritance sucks
anyway)

chimp1984 committed:
Refactoring: Move ArbitratorModule to parent package

chimp1984 committed:
Refactoring: Rename ArbitratorModule to DisputeModule

chimp1984 committed:
Add mediators to Filter

chimp1984 committed:
Add mediators to filter window

chimp1984 committed:
Use abstract DisputeResolverService as base class for ArbitratorService

- Add common base class for ArbitratorService and MediatorService

chimp1984 committed:
Fix test

chimp1984 committed:
Use abstract DisputeResolverManager as base class for ArbitratorManager

- Add common base class for ArbitratorManager and MediatorManager

chimp1984 committed:
Refactor: Move arbitratorregistration package inside register pkg

chimp1984 committed:
Refactor: Rename arbitratorregistration package to arbitrator

chimp1984 committed:
Add registration view for mediator

- With cmd+D one can open the mediator regisration in account screen.
For arbitrator its cmd+R

chimp1984 committed:
Separate pub key list for mediator (no new keys added yet)

chimp1984 committed:
Set new pubkeys for mediator registration

- Before release set new keys from maintainer who manages keys

chimp1984 committed:
Set disputes @Nullable. Add null checks

chimp1984 committed:
Remove pre v0.9 handling for supported arbitrators from offer

- We changed handling of arbitrator selection with v0.9 so the
supported arbitrators in the offer is not used anymore. As we
enforced v1.2 a while back for trading we can be sure no pre v0.9
clients are used anymore and we can remove the optional code part.

chimp1984 committed:
Remove supported arbitrators info in offer details window

- As we do not use supported arbitraors in offer anymore since v0.9
we can remove that.

chimp1984 committed:
Remove check for matching arbitrator languages

As we do not use the supported arbitratos from offer since v0.9 we can
remove that check.

chimp1984 committed:
Remove not used classes

chimp1984 committed:
Remove checks for arbitrator and mediator in offer

We do not use those fields anymore. We still need to keep the fields
not nullable as old clients have the check still.

chimp1984 committed:
Add check if sig of proto object is not empty

We got in dev testing sometimes an empty protobuf Alert. Might be
caused from protobuf copatibility issues during development but not
100% clear.
As it causes an exception and corrupted user db file we prefer to set
it to null.

chimp1984 committed:
Remove TakerSelectMediator

This is not used anymore. Currently we would get an exception in the
trade but with follow up changes we will fix that...

Mediator handling and selection will be done the same way like
arbitrator. The current mediator handling was a relict from earlier
partial support for mediators which never got completed. As still a
null check is in place we need to ensure backward compatibility.

chimp1984 committed:
Set arbitratorNodeAddresses and mediatorNodeAddresses to deprecated

We do not use arbitratorNodeAddresses and mediatorNodeAddresses anymore
but as there is a null check we still need to keep the field ans set it
to an empty arrayList.

chimp1984 committed:
Make ArbitratorSelection generic. Add MEDIATOR_ADDRESS

We want to use the same selection algorithm for mediators as for
arbitrators, so we make ArbitratorSelection generic.
We add MEDIATOR_ADDRESS as extraMap entry to TradeStatistics2 to be
able to track number of trades with specific mediators.

ExtraMap is used to add new data to existing protobuf definitions which
is supported also by not updated clients. Adding a new protobuf field
would only be supported by new clients. As mediator support is a new
feature we could add a new field but to keep it in the same style like
arbitrator we prefer to use the map here as well.

chimp1984 committed:
Refactor: Rename ArbitratorSelection to DisputeResolverSelection

chimp1984 committed:
Add mediator to OfferAvailabilityResponse and mediatorNodeAddress to OpenOffer

WIP for supporting mediator selection the same way like arbitrators.

chimp1984 committed:
Make arbitrator not nullable

We can ensure that all users are post v0.9 so we can remove the nullable
 support.

chimp1984 committed:
Add selectedMediator to OfferAvailabilityModel

Remove nullable support in ProcessOfferAvailabilityResponse as we can
ensure all clients are post v0.9

sqrrm commented on core/src/main/java/bisq/core/dispute/DisputeResolverService.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @NotNull
+    protected abstract Set<T> getCollect(List<String> bannedDisputeResolvers);
```
`getCollect()` is not a great name, `getSet()` might be better

chimp1984 replied:
Upps, was not intended. Will change it.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/account/register/DisputeResolverRegistrationView.java:
```diff
@@ -0,0 +1,270 @@
+
+        Label infoLabel = addMultilineLabel(gridPane, gridRow);
+        GridPane.setMargin(infoLabel, new Insets(Layout.TWICE_FIRST_ROW_AND_GROUP_DISTANCE, 0, 0, 0));
+        infoLabel.setText(Res.get("account.arbitratorRegistration.info.msg"));
```
This info text is not generic enough for both mediator and arbitrator

chimp1984 replied:
Ah good point. Wanted to ask @ripcurlx how it is for translators if we change key of string. I fear they will see he old key removed and need to add the new one again. That's why I left it currently, but I will add a TODO.

sqrrm commented on core/src/main/java/bisq/core/dispute/DisputeResolverManager.java:
```diff
@@ -100,32 +99,16 @@ public DisputeResolverManager(KeyRing keyRing,
     // Abstract methods
     ///////////////////////////////////////////////////////////////////////////////////////////
 
+    protected abstract List<String> getPubLeyList();
```
Should be getPub**K**eyList

chimp1984 replied:
Ups, will fix it.

sqrrm reviewed (commented):
Some minor comments

chimp1984 committed:
Refactor: Rename method

chimp1984 committed:
Add todo for using more generic keys for display strings

chimp1984 committed:
Refactor: Rename method

chimp1984 committed:
Fix wrong handling of registeredMediator

Fix copy/paste error

chimp1984 committed:
Add mediatorNodeAddress to trade

chimp1984 committed:
Handle nullable mediator in ProcessOfferAvailabilityResponse

We do not get the mediator set from old clients but we expect a not null
value so we use the DisputeResolverSelection in case it is null.
We need to pass mediatorManager and tradeStatisticsManager to the
OfferAvailabilityModel.

chimp1984 committed:
Change log level, cleanup

chimp1984 committed:
Revert changes in OfferPayload due backward compatibility issues

Because of backward compatibility issues we needed to revert the removal
of arbitratorNodeAddresses and mediatorNodeAddresses. The signature
check for the offer  would fail as an old client would send a not-empty
list but new clients would have had an empty list, so the hash
would be different and the sig check fail and we would not accept that
offer. That is the reason why we still need to support those data even
it is not used anymore.

This is one of the more tricky cases for backward compatibility issues.

This version now is tested between new and old clients and trade and
disputes work.

chimp1984 committed:
Add checks if any mediator is available

chimp1984 committed:
Cleanup classes

chimp1984 committed:
Fix test

chimp1984 committed:
Add mediator DisputeStates

Add isMediationDispute to Dispute class.

If a dispute opening gets requested we check if state is
DisputeState.NO_DISPUTE and the open mediation. If state is
DisputeState.MEDIATION_REQUESTED we open arbitration.

chimp1984 committed:
Cleanup; support isMediationDispute

chimp1984 committed:
Handle mediator data in Dispute domain

- Add getConflictResolverNodeAddress method to Dispute to resolve
arbitrator or mediator address based on isMediationDispute flag.

- Rename arbitratorPubKeyRing to conflictResolverPubKeyRing in Dispute.
We cannot rename arbitratorPubKeyRing in the protobuf definition
as it would break backward compatibility.

chimp1984 committed:
Add support for mediation in dispute domain

- Add isMediationDispute method to ChatSession
- Add isMediationDispute method to DisputeCommunicationMessage
- Add isMediationDispute to dispute id
- Refactor findDispute method
- Add null checks
- Cleanups

chimp1984 committed:
Remove impossible case

Reserved and locked funds are used for offers and trades only.

chimp1984 committed:
Fix typos

chimp1984 committed:
Handle mediator and arbitrator strings

- Work in progress of adjusting correct terms.
- Cleanups

chimp1984 committed:
Refactor: Rename arbitrator package to disputeresolvers

chimp1984 committed:
Refactor: Rename ArbitratorDisputeView classes to DisputeResolverView

chimp1984 committed:
Add support for close ticket from mediator (WIP)

In mediator case we do not create any transaction but only send the
dispute result which contains the mediators recommended payout
distribution. At teh traders we set the disputeState in the trade to
closed. This will be used in the next commits to update the trade so
that the traders get displayed the recommended payout and get asked if
they agree to that.

chimp1984 commented:
Rebased with current master.

chimp1984 commented:
I will close that branch as I run into merge issues I cannot resolve ;-(. Did some master merge and rebase locally without keeping in sync with remote and now any merge attempt fails. Was heavy refactorings with many moved and removed classes, so some complexity there screwed it up...
